### PR TITLE
✨ Recursively search for terraform hcl files

### PR DESCRIPTION
**Pull request by czunker**
Signed-off-by: Christian Zunker <christian@mondoo.com>

### CODE REVIEW

**Comment by czunker** [comment]
This works for cnquery:
```
cnquery run terraform ~/workspace/mondoo/github.com/mondoo-operator/.github -c 'terraform.providers' 
→ discover related assets for 1 asset(s)
→ resolved assets resolved-assets=1
terraform.providers: [
  0: terraform.block id = terraform.block//home/christian/workspace/mondoo/github.com/mondoo-operator/.github/terraform/gke/main.tf/31/1
  1: terraform.block id = terraform.block//home/christian/workspace/mondoo/github.com/mondoo-operator/.github/terraform/aws/provider.tf/1/1
  2: terraform.block id = terraform.block//home/christian/workspace/mondoo/github.com/mondoo-operator/.github/terraform/aws/provider.tf/5/1
  3: terraform.block id = terraform.block//home/christian/workspace/mondoo/github.com/mondoo-operator/.github/terraform/aks/versions.tf/35/1
]
```

But it seems to break `cnspec`. Have to do some more testing.

File: motor/providers/terraform/provider.go [hunk 1]
```diff
@@ -79,22 +85,36 @@ func New(tc *providers.Config) (*Provider, error) {
...
-			if err != nil {
-				return nil, errors.Wrap(err, "could not parse tfvars files")
-			}
-
-			modulesManifest, err = ParseTerraformModuleManifest(path)
+			filepath.WalkDir(path, func(path string, d fs.DirEntry, err error) error {
+				if err != nil {
+					return err
+				}
+				foundExamples := MODULE_EXAMPLES.FindString(path)
```
Comment (czunker) [inline motor/providers/terraform/provider.go:92]: I'd like to have opinions on this one. I don't think it makes sense when we scan example file like they are included in modules: `mondoo-operator/.github/terraform/aws/.terraform/modules/vpc/examples/secondary-cidr-blocks/main.tf`

Should we perhaps skip more?

File: motor/providers/terraform/provider.go [hunk 2]
```diff
@@ -79,22 +85,36 @@ func New(tc *providers.Config) (*Provider, error) {
...
+					if err != nil {
+						return errors.Wrap(err, "could not parse hcl file")
+					}
+
+					err = ReadTfVarsFromFile(path, tfVars)
+					if err != nil {
+						return errors.Wrap(err, "could not parse tfvars file")
+					}
+				} else {
+					modulesManifest, err = ParseTerraformModuleManifest(path)
```
Comment (czunker) [inline motor/providers/terraform/provider.go:109]: As I don't know which directory holds the `.terraform` subdirectory, I have to scan each directory for it. But it is no error, when it is not found.

File: motor/providers/terraform/provider.go (hunk 1 above)
Comment (imilchev) [reply motor/providers/terraform/provider.go:92]: So from what I read `examples` is a standard folder in terraform. Would there be any reason why someone would ever want to scan the content of that folder?

File: motor/providers/terraform/provider.go [hunk 3]
```diff
@@ -79,22 +85,36 @@ func New(tc *providers.Config) (*Provider, error) {
...
-			}
-
-			modulesManifest, err = ParseTerraformModuleManifest(path)
+			filepath.WalkDir(path, func(path string, d fs.DirEntry, err error) error {
+				if err != nil {
+					return err
+				}
+				foundExamples := MODULE_EXAMPLES.FindString(path)
+				if foundExamples != "" {
+					log.Debug().Str("path", path).Msg("ignoring terraform module example")
```
Comment (imilchev) [inline motor/providers/terraform/provider.go]: if we are ignoring the folder I would log a warning message so it is explicitly clear we skipped that folder

File: motor/providers/terraform/provider.go (hunk 1 above)
Comment (czunker) [reply motor/providers/terraform/provider.go:92]: I don't know why someone would scan these. When they use the examples, the code will end up in a place where is relevant and scanned.

File: motor/providers/terraform/provider.go (hunk 3 above)
Comment (czunker) [reply motor/providers/terraform/provider.go]: Changed.

**Review by imilchev — APPROVED** [review]
The code looks good to me but I don't know too much about TF. I tried it out with a few configs and it seems to work. Can't come up with any edge-cases though

**Pull request merged**